### PR TITLE
Milestone3 - Opshin Edge case Identification

### PR DESCRIPTION
**Pull request by SuganyaAK**
I have prepared this report as per the catalyst acceptance criteria
**Quantification of Edge Cases:** The list includes a specified number of edge cases (e.g., a minimum of 5 unique edge cases), each thoroughly described and categorized.
**Validation by Development:** The list of edge cases and proposed resolutions is reviewed and validated by OpShin team

### CODE REVIEW

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+```
+
+## 6. Uncovered Scenarios in `Union` Type Behavior
+
+- **File Reference**: `opshin/tests/test_Unions.py`
+- **Edge Cases**:
+  The following edge cases were not included in the test suite.
+
+  1.Duplicate entries in `Unions` give compiler errors, but duplicate entries in nested `Unions` don't.
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:157]: Let's list the edge cases first and then proceed to the recommendations section. Similar to this:

Edge Cases:
1. ...
2. ...
3. ...

Recommendations:
1. ...
2. ...
3. ...

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+### Recommendation:
+
+We recommend to sort the `Union` entries in an unambiguous manner within the `union_types()` function in `type_inference.py`.
+
+## 7. Explicit testing for edge cases in bitmap and fractions
+
+- **File Reference**: `opshin/tests/test_bitmap.py`,`opshin/tests/test_fractions.py`
+- **Edge Case**:
+
+`test_fractions.py`: The existing sample cases exclude scenarios where the denominator is zero.
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:225]:
```suggestion
`test_fractions.py`: The existing sample cases does not include scenarios where the denominator is zero.
```

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+`test_bitmap.py`: Similar oversight with excluding empty bytestrings and negative values,these should also be tested to ensure robustness.
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:227]: No, we should not try and figure out why the issue is present!
```suggestion
`test_bitmap.py`: Similarly there are no test cases covering empty bytestring inputs and negative input values (e.g. creating a bitmap with a negative length).
```

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+### Recommendation:
+
+Add explicit test cases for
+
+- Denominator = 0 in fractions (to check handling of division by zero).
+- Empty bytestrings and negative index values in bitmap operations.
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:234]:
```suggestion
- Empty bytestrings and negative index/length values in bitmap operations.
```

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+## 8. Index type not tested for different types
+
+- **File Reference**: `opshin/tests/test_stdlib.py`
+
+- **Edge Case**:
+
+Attribute of listtype "index" is only tested for integers, missing other types which leads to the following edge case that wasn't tested.
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:242]:
```suggestion
Attribute of listtype "index" is only tested for integers, missing other types, for example:
```

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+
+def validator(a: Anything, b: Anything) -> int:
+   l: List[Anything] = [a, b]
+
+   return l.index(b)
+```
+
+### Recommendation:
+
+In the `index` method, defined in `ListType.attribute()` in `type_impls.py`, change the check to `EqualsData(transform_output_map(itemType)(x), transform_output_map(itemType)(HeadList(xs)))`.
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:255]: I don't think we should explicitly dictate code; we should just describe the gist of the recommended changes.

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+
+## 9. No test cases for `CONSTR_ID` attribute.
+
+- **Edge Case**:
+
+No test cases for analysing the behaviour of `CONSTR_ID` attribute.
+The following edge cases are valid OpShin, but isn't consistent with how attributes are exposed of regular `Union`s (they must exist on each subtype), and can lead to unexpected runtime errors.
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:262]:
```suggestion
The following edge cases are valid in OpShin, but isn't consistent with how attributes are used in other types (e.g. `Union`s).
```

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+
+def validator(l: List[Anything]) -> int:
+    return l[0].CONSTR_ID
+```
+
+```python
+from opshin.prelude import
+
+def validator(u: Union[int, bytes]) -> int:
+return u.CONSTR_ID
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:275]:
```suggestion
    return u.CONSTR_ID
```

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+
+### Recommendation:
+
+In the `index` method, defined in `ListType.attribute()` in `type_impls.py`, change the check to `EqualsData(transform_output_map(itemType)(x), transform_output_map(itemType)(HeadList(xs)))`.
+
+## 9. No test cases for `CONSTR_ID` attribute.
+
+- **Edge Case**:
+
+No test cases for analysing the behaviour of `CONSTR_ID` attribute.
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:261]:
```suggestion
No test cases for validating the computation of `CONSTR_ID` attributes.
```

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+from opshin.prelude import
+
+def validator(u: Union[int, bytes]) -> int:
+return u.CONSTR_ID
+```
+
+### Recommendation:
+
+1. Remove the `CONSTR_ID` attribute for `Anything`.
+2. Don't expose the `CONSTR_ID` attribute of `Union`s which contain some non-`ConstrData` types.
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:281]:
```suggestion
2. Do not expose the `CONSTR_ID` attribute of `Union`s which contain some non-`ConstrData` types.
```

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+**Milestone Title** : Milestone3 - Edge Case Identification  
+**Submission Date** : 23-May-2025  
+**Audit Team** : Anastasia Labs  
+**Repository** : [Opshin Repository](https://github.com/OpShin/opshin)
+
+## Executive Summary
+
+This report documents edge cases in OpShin smart contract development identified through manual code review. It provides actionable strategies to resolve these issues that can be applied to future iterations of the language.
+
+### Scope
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:13]:
```suggestion
## Scope
```

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+
+We took the following approach to list the edge cases:
+
+- Cross-referencing unit test files under `opshin/tests` against language specifications to uncover untested functionality.
+
+- Incorporating documented edge cases from Milestone 2 to highlight high-risk scenarios lacking test coverage.
+
+# Edge cases
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:21]:
```suggestion
## Edge cases
```

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+
+## 1. Untested builtin `len()`
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:23]:
```suggestion
### 1. Untested builtin `len()`
```

Comment (mpetruska) [reply catalyst_docs/milestone3/milestone3-edgecases.md:23]: And so on... we should have a consistent document structure (# Title should appear only once):

```
# OpShin Edge Case Analysis Report
## Summary
## Scope
## Edge cases
### Finding 1...
### Finding 2...
...
### Final remarks/notes/administation, etc.
```

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+def validator(u: Union[int, bytes]) -> int:
+return u.CONSTR_ID
+```
+
+### Recommendation:
+
+1. Remove the `CONSTR_ID` attribute for `Anything`.
+2. Don't expose the `CONSTR_ID` attribute of `Union`s which contain some non-`ConstrData` types.
+
+# Summary
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:283]: There is already a summary section on top, this is definitely not needed. Please merge the contents into the one on top.

File: catalyst_docs/milestone3/milestone3-edgecases.md [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+
+This report identifies and analyzes 9 edge cases across the OpShin codebase and test suite.
+Each case highlights a gap in either functionality coverage or robustness in the current implementation of the language.
+
+For each edge case, we have proposed actionable recommendations, such as test coverage expansion,and improvements to type checking logic.
+Implementing these changes will improve OpShin’s reliability, developer experience, and readiness for production smart contract development.
+
+# Validation by OpShin Team
```
Comment (mpetruska) [inline catalyst_docs/milestone3/milestone3-edgecases.md:291]:
```suggestion
## Validation by OpShin Team
```

Comment (mpetruska) [reply catalyst_docs/milestone3/milestone3-edgecases.md:291]: I don't think this section is needed at all.

**Review by mpetruska — CHANGES_REQUESTED** [review]
Please look at my comments/suggestions.

File: catalyst_docs/milestone3/milestone3-edgecases.md (hunk 14 above)
Comment (SuganyaAK) [reply catalyst_docs/milestone3/milestone3-edgecases.md:285]: I have merged it with the executive summary

File: catalyst_docs/milestone3/milestone3-edgecases.md (hunk 6 above)
Comment (SuganyaAK) [reply catalyst_docs/milestone3/milestone3-edgecases.md:257]: changed to 
In the `index` method defined in `ListType.attribute()` in `type_impls.py`, update the check to accommodate other primitive types, and add corresponding test cases to ensure coverage.

File: catalyst_docs/milestone3/milestone3-edgecases.md (hunk 15 above)
Comment (SuganyaAK) [reply catalyst_docs/milestone3/milestone3-edgecases.md:293]: Removed

**Comment by SuganyaAK** [comment]
> Nice work, only some minor stylistic suggestions.

Thank you Mark, Resolved all comments.
